core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -178,84 +185,104 @@ public AllocateUnassignedDecision makeAllocationDecision(final ShardRouting unas
-                    buildNodeDecisions(nodesToForceAllocate, explain));
-            } else if (nodesToForceAllocate.throttleNodeShards.isEmpty() == false) {
+                node = nodeShardState.getNode();
+                allocationId = nodeShardState.allocationId();
+            } else if (nodesToAllocate.throttleNodeShards.isEmpty() == false) {
                 logger.debug("[{}][{}]: throttling allocation [{}] to [{}] on forced primary allocation",
-                             unassignedShard.index(), unassignedShard.id(), unassignedShard, nodesToForceAllocate.throttleNodeShards);
-                return AllocateUnassignedDecision.throttle(
-                    explain ? "allocation throttled as all nodes to which the shard may be force allocated are busy with other recoveries" : null,
-                    buildNodeDecisions(nodesToForceAllocate, explain));
+                             unassignedShard.index(), unassignedShard.id(), unassignedShard, nodesToAllocate.throttleNodeShards);
+                throttled = true;
             } else {
                 logger.debug("[{}][{}]: forced primary allocation denied [{}]",
                              unassignedShard.index(), unassignedShard.id(), unassignedShard);
-                return AllocateUnassignedDecision.no(AllocationStatus.DECIDERS_NO,
-                    explain ? "all nodes that hold a valid shard copy returned a NO decision, and force allocation is not permitted" : null,
-                    buildNodeDecisions(nodesToForceAllocate, explain));
             }
         } else {
             // we are throttling this, since we are allowed to allocate to this node but there are enough allocations
             // taking place on the node currently, ignore it for now
             logger.debug("[{}][{}]: throttling allocation [{}] to [{}] on primary allocation",
                          unassignedShard.index(), unassignedShard.id(), unassignedShard, nodesToAllocate.throttleNodeShards);
-            return AllocateUnassignedDecision.throttle(
-                explain ? "allocation throttled as all nodes to which the shard may be allocated are busy with other recoveries" : null,
-                buildNodeDecisions(nodesToAllocate, explain));
+            throttled = true;
+        }
+
+        Map<String, NodeAllocationResult> nodeResults = null;
+        if (explain) {
+            nodeResults = buildNodeDecisions(nodesToAllocate, inSyncAllocationIds);
+        }
+        if (allocation.hasPendingAsyncFetch()) {
+            return AllocateUnassignedDecision.no(AllocationStatus.FETCHING_SHARD_DATA, nodeResults);
+        } else if (node != null) {
+            return AllocateUnassignedDecision.yes(node, allocationId, nodeResults, false);
+        } else if (throttled) {
+            return AllocateUnassignedDecision.throttle(nodeResults);
+        } else {
+            return AllocateUnassignedDecision.no(AllocationStatus.DECIDERS_NO, nodeResults, true);
```

Comment:
The reason I went with this way is because the `AllocateUnassignedDecision` has a lot of constructor params, so doing it this way enables passing as few params as needed to the static method to create the `AllocateUnassignedDecision`, and also makes it more clear what type of decision object is being created.  That was the main reason for doing it like this.  WDYT?